File: 백준/Gold/12764. 싸지방에 간 준하/싸지방에 간 준하.cc
#include <iostream>
#include <vector>
#include <queue>
#include <stack>
#include <map>
#include <set>
#include <string>
#include <algorithm>
#include <memory.h>
#include <cmath>

using namespace std;
#define IAMFAST ios_base::sync_with_stdio(false);cin.tie(0);
typedef long long ll;
typedef pair<int, int> pii;

int n;
vector<pii> v;
priority_queue<int, vector<int>, greater<>> blankSeat;
priority_queue<pii, vector<pii>, greater<>> pq;
int cnt[100001];
int seatNum = 1;

void INPUT()
{
   IAMFAST
   cin >> n;
   for (int i = 0; i < n; i++)
   {
      int s, e;
      cin >> s >> e;
      v.emplace_back(s, e);
   }

}

void solution()
{
   sort(v.begin(), v.end());

   pq.emplace(v[0].second, seatNum);
   cnt[seatNum]++;
   for (int i = 1; i < v.size(); i++)
   {
      auto &[s, e] = v[i];

      while (!pq.empty() && s > pq.top().first)
      {
         blankSeat.emplace(pq.top().second);
         pq.pop();
      }

      if (blankSeat.empty())
      {
         pq.emplace(e, ++seatNum);
//         cout << s << " ~ " << e << " => " << seatNum << '\n';
         cnt[seatNum]++;
      }
      else
      {
         pq.emplace(e, blankSeat.top());
         cnt[blankSeat.top()]++;
//         cout << s << " ~ " << e << " => " << blankSeat.top() << '\n';
         blankSeat.pop();
      }
   }

   cout << seatNum << '\n';
   for (int i = 1; i <= seatNum; i++)
      cout << cnt[i] << " ";

}

int main()
{
   INPUT();
   solution();
}

File: 백준/Gold/12764. 싸지방에 간 준하/싸지방에 간 준하.java
import java.io.*;
import java.util.*;

public class Main {

  static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
  static BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
  static StringTokenizer st;

  public static class pii implements Comparable<pii> {

    int first;
    int second;

    pii(int first, int second) {
      this.first = first;
      this.second = second;
    }

    ;

    @Override
    public int compareTo(pii a) {
      return this.first - a.first;
    }
  }

  static int n;
  static ArrayList<pii> v = new ArrayList<>();
  static PriorityQueue<Integer> blankSeat = new PriorityQueue<>();
  static PriorityQueue<pii> pq = new PriorityQueue<>();
  static int[] cnt = new int[100001];
  static int seatNum = 1;

  public static void input() throws IOException {
    n = Integer.parseInt(br.readLine());
    for (int i = 0; i < n; i++) {
      st = new StringTokenizer(br.readLine());
      int s = Integer.parseInt(st.nextToken());
      int e = Integer.parseInt(st.nextToken());
      v.add(new pii(s, e));
    }
  }

  public static void solution() throws IOException {
    Collections.sort(v);

    pq.offer(new pii(v.get(0).second, seatNum));
    cnt[seatNum]++;
    for (int i = 1; i < v.size(); i++) {
      int s = v.get(i).first;
      int e = v.get(i).second;

      while (!pq.isEmpty() && s > pq.element().first) {
        blankSeat.offer(pq.element().second);
        pq.remove();
      }

      if (blankSeat.isEmpty())
      {
        pq.offer(new pii(e, ++seatNum));
        cnt[seatNum]++;
      }
      else
      {
        pq.offer(new pii(e, blankSeat.element()));
        cnt[blankSeat.element()]++;
        blankSeat.remove();
      }
    }

    bw.write(seatNum + "\n");
    for (int i = 1; i <= seatNum; i++)
      bw.write(cnt[i] + " ");
    
    br.close();
    bw.flush();
    bw.close();
  }

  public static void main(String[] args) throws IOException {

    input();
    solution();

  }
}
